src/stdlib/data.rs

```diff
@@ -0,0 +1,382 @@
+    map_indexed_func,
+    |ctx, list, func| {
+        expect_iterable!(list => iter);
+        expect_func!(func => func);
+        let func = func.clone();
+        let i = Rc::new(RefCell::new(0));
+        Ok(FenderIterator {
+            next: Rc::new(move |e| {
+                let elem = (iter.next)(e)?;
+                let mut borrow = i.borrow_mut();
+                let val = e.call(&func, vec![FenderValue::Int(*borrow).into(), elem])?;
+                *borrow += 1;
+                Ok(val)
+            }),
+            has_next: iter.has_next.clone(),
+        }
+        .into())
+    }
+);
+
+fndr_native_func!(
+    /// Splits a string given a certain substring or character
+    split_func,
+    |ctx, str, delim| {
+        use FenderValue::*;
+        type_match!(str, delim {
+            (String(str), String(delim)) => {
+                let list = str
+                    .split(&**delim)
+                    .map(|v| FenderValue::make_string(v.to_string()).into())
+                    .collect();
+                Ok(FenderValue::make_list(list).into())
+            },
+            (String(str), Char(c)) => {
+                let list = str
+                    .split(c)
+                    .map(|v| FenderValue::make_string(v.to_string()).into())
+                    .collect();
+                Ok(FenderValue::make_list(list).into())
+            }
+        })
+    }
+);
+
+fndr_native_func!(
+    /// Converts a string to uppercase
+    upper_func,
+    |ctx, str| {
+        use FenderValue::*;
+        type_match!(str {
+            (String(s)) => Ok(FenderValue::make_string(s.to_uppercase()).into()),
+            (Char(c)) => Ok(FenderValue::Char(c.to_ascii_uppercase()).into())
+        })
+    }
+);
+
+fndr_native_func!(
+    /// Converts a string to lowercase
+    lower_func,
+    |ctx, str| {
+        use FenderValue::*;
+        type_match!(str {
+            (String(s)) => Ok(FenderValue::make_string(s.to_lowercase()).into()),
+            (Char(c)) => Ok(FenderValue::Char(c.to_ascii_lowercase()).into())
+        })
+    }
+);
+
+fndr_native_func!(
+    /// Calculates the sum of an iterable
+    sum_func,
+    |ctx, list| {
+        use FenderValue::*;
+        expect_iterable!(list => iter);
+        let mut sum = Int(0).into();
+        while *(iter.has_next)(ctx)? == Bool(true) {
+            let elem = (iter.next)(ctx)?;
+            sum = FenderBinaryOperator::Add.apply_2(&sum, &elem);
+        }
+        Ok(sum)
+    }
+);
+
+fndr_native_func!(
+    /// Calculates the sum of an iterable
+    product_func,
+    |ctx, list| {
+        use FenderValue::*;
+        expect_iterable!(list => iter);
+        let mut sum = Int(1).into();
+        while *(iter.has_next)(ctx)? == Bool(true) {
+            let elem = (iter.next)(ctx)?;
+            sum = FenderBinaryOperator::Mul.apply_2(&sum, &elem);
+        }
+        Ok(sum)
+    }
+);
+
+fndr_native_func!(
+    /// Remove elements according to a filtering function
```

Comment:
wording of this makes it sound like you are selecting the elements to be removed, maybe the following sounds better
```suggestion
    /// Remove elements that return `false` when passed to `pred`
```